### PR TITLE
No user dependency + requiered fields in Cust table

### PR DESCRIPTION
**Pull request by mayabrandi**
Make links between customer contacts and user table.
Make invoice_address and invoice_reference requiered in Customer model

How to test:

1 Install the no_user_dependency_on branch on rasta and on clinical-db. STAGE!!!
2 Upgrade database with scripts/make_customer_information_mandatory.sql script
3 Go to https://clinical-api-stage.scilifelab.se/admin/customer/ and try to add a new customer.
3.1 See that Invoice adress and Invoice reference now are mandatory.
3.2 See that Primary Contact, Invoice Contact and Delivery Contact now are drop downs with names defined in the user view.

### CODE REVIEW

File: cg/store/api/find.py [hunk 1]
```diff
@@ -29,6 +29,10 @@ def customer_by_id(self, id_: int) -> models.Customer:
     def user(self, email: str) -> models.User:
         """Fetch a user from the store."""
         return self.User.query.filter_by(email=email).first()
+
```
Comment (patrikgrenfeldt) [inline cg/store/api/find.py]: please delete method if your not using it

File: cg/store/models.py [hunk 2]
```diff
@@ -64,8 +64,8 @@ class Customer(Model):
     project_account_ki = Column(types.String(32))
     project_account_kth = Column(types.String(32))
     organisation_number = Column(types.String(32))
-    invoice_address = Column(types.Text)
-    invoice_reference = Column(types.String(32))
+    invoice_address = Column(types.Text, nullable=False)
```
Comment (patrikgrenfeldt) [inline cg/store/models.py]: Please write script to update database to reflect the model changes

File: cg/store/api/find.py (hunk 1 above)
Comment (mayabrandi) [reply cg/store/api/find.py]: I didnt write this function so im not sure its not being used in some other code.

File: cg/store/models.py (hunk 2 above)
Comment (mayabrandi) [reply cg/store/models.py]: Emilia and Vallteri are taking care of adding missing data to those fields.

File: cg/store/api/find.py (hunk 1 above)
Comment (patrikgrenfeldt) [reply cg/store/api/find.py]: The '+' signs in github indicates that the 'user_by_cust' got added in your commit, right? And if you don't call it, it is no longer needed, right?

**Comment by mayabrandi** [comment]
Yes absolutely! Sorry. When I read your comment on git hub It looked as if you were referring to the user function! I was certain I had removed the user_by_cust already but obviously I hadn’t! Done now though! I pushed it a little while ago.

/M

File: cg/store/models.py (hunk 2 above)
Comment (patrikgrenfeldt) [reply cg/store/models.py]: Yes, I'm talking about the database structure, not the data

**Comment by mayabrandi** [comment]
Aha!! :) But no need for a script to do that? Or what do you mean? Its its done in just a click using the sequel pro app eg. Or I misunderstand you?

File: cg/store/models.py (hunk 2 above)
Comment (patrikgrenfeldt) [reply cg/store/models.py]: I suggest adding a script, something like make_invoice_information_mandatory.sql, in the scripts folder doing something like: https://stackoverflow.com/questions/689746/altering-a-column-null-to-not-null

**Comment by mayabrandi** [comment]
But why? When its done by just unclicking ”tillåt Null” in Sequel Pro?

**Comment by ingkebil** [comment]
If we ever want to create unit tests in regards to databases, we need set up (and tear down) scripts that modify the db structure. So creating them now creates a good habit.

But the main reason now is: The scripts are reproducible and we can trace back if we e.g. put in the wrong type into production. Doing something manual does not have this advantage.

**Comment by mayabrandi** [comment]
Ok! Thnks for the explanation!

/M

**Comment by mayabrandi** [comment]
I am hesitant to change the name of the columns in the database (although I agree on that the column names should have been more descriptive when they were defined). At least I think we need to go through all code that might be referring to those columns so that nothing crashes. And that should be as part of a separate PR.

Well lets discuss wether it should be a reference or not. And the same as with your suggestion as above... We should probably go through code and make sure nothing is crashing by changing the field? Or maybe thats not a risk? It should be in a separate PR any how.

**Comment by mayabrandi** [comment]
Ill need a little help on this. Ill make the script but Im not sure how to test it, or even run it!

**Comment by patrikgrenfeldt** [comment]
@mayabrandi looking at the testresults here: https://travis-ci.org/Clinical-Genomics/cg/builds/444950637?utm_source=github_status&utm_medium=notification, it seems to me that we should add the new NOT NULL constrained parameters when creating customers in the tests that fail, right?

**Comment by patrikgrenfeldt** [comment]
Did you remember to create your environment as a python environment? I followed this guide: [Guide](https://uoa-eresearch.github.io/eresearch-cookbook/recipe/2014/11/20/conda/) and then the tests run just fine
Can you please make sure that this is not the problem before we sit down together?

**Comment by mayabrandi** [comment]
Is this one okey now? @patrikgrenfeldt Can we put it in to production now?

File: scripts/make_customer_information_mandatory.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
+UPDATE customer SET invoice_address="" WHERE invoice_address IS NULL;
+UPDATE customer SET invoice_reference="" WHERE invoice_reference IS NULL;
+ALTER TABLE customer MODIFY COLUMN invoice_address text NOT NULL;
+ALTER TABLE customer MODIFY COLUMN invoice_reference varchar(32) NOT NULL;
+
+ALTER TABLE customer ADD COLUMN delivery_contact_id int NULL;	
+Update customer SET delivery_contact_id = ( SELECT id FROM user Where email = customer.delivery_contact ) Where delivery_contact_id is NULL;	
+ALTER TABLE customer ADD CONSTRAINT `delivery_contact_ibfk_1` FOREIGN KEY (`delivery_contact_id`) REFERENCES `user` (`id`);	
+ALTER TABLE customer DROP COLUMN delivery_contact
```
Comment (patrikgrenfeldt) [inline scripts/make_customer_information_mandatory.sql]:
```suggestion
ALTER TABLE customer DROP COLUMN delivery_contact;
```

File: scripts/make_customer_information_mandatory.sql [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+ALTER TABLE customer ADD COLUMN delivery_contact_id int NULL;	
+Update customer SET delivery_contact_id = ( SELECT id FROM user Where email = customer.delivery_contact ) Where delivery_contact_id is NULL;	
+ALTER TABLE customer ADD CONSTRAINT `delivery_contact_ibfk_1` FOREIGN KEY (`delivery_contact_id`) REFERENCES `user` (`id`);	
+ALTER TABLE customer DROP COLUMN delivery_contact	
+
+ALTER TABLE customer ADD COLUMN invoice_contact_id int NULL;	
+Update customer SET invoice_contact_id = ( SELECT id FROM user Where email = customer.invoice_contact ) Where invoice_contact_id is NULL;	
+ALTER TABLE customer ADD CONSTRAINT `invoice_contact_ibfk_1` FOREIGN KEY (`invoice_contact_id`) REFERENCES `user` (`id`);	
+ALTER TABLE customer DROP COLUMN invoice_contact
```
Comment (patrikgrenfeldt) [inline scripts/make_customer_information_mandatory.sql]:
```suggestion
ALTER TABLE customer DROP COLUMN invoice_contact;
```

File: scripts/make_customer_information_mandatory.sql [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+ALTER TABLE customer ADD COLUMN invoice_contact_id int NULL;	
+Update customer SET invoice_contact_id = ( SELECT id FROM user Where email = customer.invoice_contact ) Where invoice_contact_id is NULL;	
+ALTER TABLE customer ADD CONSTRAINT `invoice_contact_ibfk_1` FOREIGN KEY (`invoice_contact_id`) REFERENCES `user` (`id`);	
+ALTER TABLE customer DROP COLUMN invoice_contact	
+
+ALTER TABLE customer ADD COLUMN primary_contact_id int NULL;	
+Update customer SET primary_contact_id = ( SELECT id FROM user Where email = customer.primary_contact ) Where primary_contact_id is NULL;	
+ALTER TABLE customer ADD CONSTRAINT `primary_contact_ibfk_1` FOREIGN KEY (`primary_contact_id`) REFERENCES `user` (`id`);	
+ALTER TABLE customer DROP COLUMN primary_contact
```
Comment (patrikgrenfeldt) [inline scripts/make_customer_information_mandatory.sql]:
```suggestion
ALTER TABLE customer DROP COLUMN primary_contact;
```

File: cg/meta/invoice.py [hunk 6]
```diff
@@ -12,27 +12,29 @@ def __init__(self, db: Store, lims_api: lims.LimsAPI, invoice_obj: models.Invoic
...
             self.log.append(msg)
             return None
+
         contact = {
-                'name': contact_user.name,
-                'email': contact_user.email,
-                'customer_name': contact_customer.name,
-                'reference': contact_customer.invoice_reference,
-                'address': contact_customer.invoice_address,
+                'name': user.name if user else '',
```
Comment (patrikgrenfeldt) [inline cg/meta/invoice.py]:
```suggestion
                'name': user.name,
```
Om vi saknar user så borde vi redan hoppat ur på rad 28?

File: cg/cli/add.py [hunk 7]
```diff
@@ -21,21 +21,27 @@ def add(context):
 @click.option('-cg', '--customer-group', 'customer_group_id', required=False,
               help='internal ID for the customer group of the customer, a new group will be '
                    'created if left out')
+@click.option('-ia', '--invoice-address', 'invoice_address', required=True,
+              help='Post adress to invoice bla bla....')
+@click.option('-ir', '--invoice-reference', 'invoice_reference', required=True,
+              help='bla bla....')
```
Comment (patrikgrenfeldt) [inline cg/cli/add.py]: this does not help ;)

**Comment by patrikgrenfeldt** [comment]
Database changes: 👍 
Functionality: 👍 
--help message: could be improved
Test coverage: could be improved

**Comment by mayabrandi** [comment]
Thanks Patrik! Ill look at the help message!

File: cg/store/api/find.py (hunk 1 above)
Comment (patrikgrenfeldt) [reply cg/store/api/find.py]: removed, great

**Comment by patrikgrenfeldt** [comment]
@mayabrandi, could you have a look at the conflicts in this PR?

File: cg/server/admin.py [hunk 8]
```diff
@@ -7,7 +7,8 @@
 
 
 class BaseView(ModelView):
-
+    can_export = True
```
Comment (patrikgrenfeldt) [inline cg/server/admin.py]: I see that we have added this in the BaseView, reason?

Comment (patrikgrenfeldt) [reply cg/server/admin.py]: I got "Tablib dependency not installed. " in all views I tried, some dependency missing?

File: cg/server/admin.py [hunk 9]
```diff
@@ -21,13 +22,11 @@ class CustomerView(BaseView):
     column_exclude_list = [
         'agreement_date',
         'organisation_number',
-        'invoice_address',
-        'delivery_contact',
-        'invoice_contact',
+        'invoice_address'
```
Comment (patrikgrenfeldt) [inline cg/server/admin.py]: Why is the invoice_address excluded?

File: cg/server/admin.py (hunk 8 above)
Comment (patrikgrenfeldt) [reply cg/server/admin.py]: ok, gone

**Pull request merged**